app/src/org/commcare/android/database/global/GlobalDatabaseUpgrader.java

```diff
@@ -78,4 +91,40 @@ private boolean upgradeTwoThree(SQLiteDatabase db) {
             db.endTransaction();
         }
     }
+
+    /**
+     * Prior to multiple application seating, the FormsProvider and the InstanceProvider were both
+     * using one global database for all forms/instances. Now that we can have multiple apps
+     * installed at once, we need to have one forms db and one instances db per app. This method
+     * performs the necessary one-time migration from a global db that still exists on a device
+     * to the new per-app system
+     */
+    private boolean upgradeProviderDb(SQLiteDatabase db, ProviderUtils.ProviderType type) {
+        File oldDbFile = CommCareApplication._().getDatabasePath(type.getOldDbName());
+        if (oldDbFile.exists()) {
+            File newDbFile = CommCareApplication._().getDatabasePath(
+                    ProviderUtils.getProviderDbName(type, getInstalledAppRecord(c, db).getApplicationId()));
+            if (!oldDbFile.renameTo(newDbFile)) {
```

Comment:
That makes sense essentially, although I'm not seeing a clear mechanism to revert/rollback in SqlLiteDatabase. Would it maybe make sense to just tell them "You need to wipe your data and restart commcare" like you suggested below? There's a small chance that just restarting commcare will fix whatever weirdness is going on causing it to fail, but if it doesn't then they're going to just be stuck in this loop of restarting and re-trying.